ftfy/bad_codecs/utf8_variants.py

```diff
@@ -46,8 +46,29 @@
 import codecs
 
 NAME = 'utf-8-variants'
-# This regular expression matches all possible six-byte CESU-8 sequences.
-CESU8_RE = re.compile(b'\xed[\xa0-\xaf][\x80-\xbf]\xed[\xb0-\xbf][\x80-\xbf]')
+
+# This regular expression matches all possible six-byte CESU-8 sequences,
+# plus truncations of them at the end of the string.
+CESU8_EXPR = (
+    b'(\xed[\xa0-\xaf][\x80-\xbf]\xed[\xb0-\xbf][\x80-\xbf]'
+    b'|\xed([\xa0-\xaf]([\x80-\xbf](\xed[\xb0-\xbf]?)?)?)?$'
+    b')'
+)
+CESU8_RE = re.compile(CESU8_EXPR)
+
+# This expression matches isolated surrogate characters that aren't
+# CESU-8, which have to be handled carefully on Python 2.
+SURROGATE_EXPR = (b'(\xed[\xa0-\xbf][\x80-\xbf])')
+
+# This expression matches the Java encoding of U+0. We don't need to check
+# for its truncated version, because that will get passed on to the standard
+# UTF-8 decoder, which will agree that we need to see more bytes.
```

Comment:
You know, I wonder if this is actually true.

Upon seeing byte C0, the UTF-8 decoder could look a bit skeptical and ask to see more bytes. Or it could just say "WTF is this" and start raising errors right away, because there is no legitimate use for byte C0.